### PR TITLE
New parser library

### PR DESCRIPTION
**Pull request by chadell**
Goal: Use new `circuit-maintenance-parser` library, with new methods to simply client usage and reorg code to simplify and increase test coverage

# Changes

- [x] Adopt the new `circuit-maintenance-parser` init methods, simplifying the logic to send the email data directly to the library. To homogenize the usage, in `GmailSources` we move to the `raw` data. This also allowed the reuse the `process_email` between `EmailSources`. This also removed a lot of logic regarding email manipulation.
- [x] Revisit the logic in `handler.py` to make it more readable and add new related testing. There are some changes on return objects in some functions (for instance, the `process_raw_notification` now creates the `RawNotification` even the provider _type is not supported by the parser library)
- [x] Check `Raw` visualisation and storage for `RawNotification` because now emails are fully dumped
- [x] Use `Stamp` from the parser `Maintenance` to set the `RawNotification.date`

I tried to simplify the logic as much as possible, but any suggestion on it will be more than welcome!

### CODE REVIEW

File: nautobot_circuit_maintenance/handle_notifications/handler.py
Comment (glennmatthews) [inline nautobot_circuit_maintenance/handle_notifications/handler.py:190]: This feels like it should be `log_failure` to me - providers shouldn't be raising any Exceptions other than ProviderError, right?

File: nautobot_circuit_maintenance/handle_notifications/handler.py
Comment (glennmatthews) [inline nautobot_circuit_maintenance/handle_notifications/handler.py:186]: This should probably at least be `log_warning`, and I could make an argument for `log_failure`. In integrating this plugin into an automated workflow (with daily or hourly runs of the importing job) we need to make it easy to automatically detect and flag parsing failures for telemetry, dashboards, etc.

File: nautobot_circuit_maintenance/handle_notifications/handler.py
Comment (glennmatthews) [inline nautobot_circuit_maintenance/handle_notifications/handler.py:196]:
```suggestion
def create_raw_notification(logger: Job, notification: MaintenanceNotification, provider: Provider):
```

File: nautobot_circuit_maintenance/handle_notifications/handler.py
Comment (glennmatthews) [inline nautobot_circuit_maintenance/handle_notifications/handler.py:239]:
```suggestion
    raw_entry = create_raw_notification(logger, notification, provider)
```

File: nautobot_circuit_maintenance/handle_notifications/handler.py
Comment (glennmatthews) [inline nautobot_circuit_maintenance/handle_notifications/handler.py:225]:
```suggestion
def process_raw_notification(logger: Job, notification: MaintenanceNotification) -> Optional[UUID]:
```

File: nautobot_circuit_maintenance/tests/test_handler.py
Comment (glennmatthews) [inline nautobot_circuit_maintenance/tests/test_handler.py:207]: Why did this change? Are we now successfully processing the raw notification even though it's from an unknown provider_type?

Comment (chadell) [reply nautobot_circuit_maintenance/tests/test_handler.py:207]: I'm still thinking on the final logic, this was just a first try on the change, but the `process_raw_notification` now creates the `RawNotification` even it can't be parsed (in this case because the parser library has no `provider`) 
Obviously this is not the final word, still in `draft` mode

File: nautobot_circuit_maintenance/handle_notifications/handler.py
Comment (chadell) [reply nautobot_circuit_maintenance/handle_notifications/handler.py:196]: it took me 30 seconds to spot the typo 😄

File: nautobot_circuit_maintenance/models.py
Comment (glennmatthews) [inline nautobot_circuit_maintenance/models.py:19]: Misleading name? We can configure a higher value than this if desired, right? Perhaps `DEFAULT_RAW_NOTIFICATION_SIZE` would be better?

This only applies to the RawNotification, right? The CircuitMaintenance will still store the full text of the email?

File: nautobot_circuit_maintenance/handle_notifications/handler.py
Comment (glennmatthews) [inline nautobot_circuit_maintenance/handle_notifications/handler.py:27]: What's the type of `parser_maintenance` now?

Same question on each function that has this as an argument below.

File: nautobot_circuit_maintenance/handle_notifications/handler.py
Comment (glennmatthews) [inline nautobot_circuit_maintenance/handle_notifications/handler.py:256]: Can we do these updates atomically as part of `create_or_update_circuit_maintenance` instead of as a separate step? I have an extension plugin that's listening for the `post_save` signal for `CircuitMaintenance` and I think with these changes it'll be triggered twice instead of once.

File: nautobot_circuit_maintenance/tests/test_handler.py
Comment (glennmatthews) [inline nautobot_circuit_maintenance/tests/test_handler.py:265]:
```suggestion
        self.assertIsNone(parser_maintenances)
```

File: nautobot_circuit_maintenance/tests/test_handler.py
Comment (glennmatthews) [inline nautobot_circuit_maintenance/tests/test_handler.py:275]:
```suggestion
        self.assertIsNone(parser_maintenances)
```

File: nautobot_circuit_maintenance/tests/test_sources.py
Comment (glennmatthews) [inline nautobot_circuit_maintenance/tests/test_sources.py:194]:
```suggestion
        self.assertIsNone(provider_type)
```

File: nautobot_circuit_maintenance/handle_notifications/handler.py
Comment (chadell) [reply nautobot_circuit_maintenance/handle_notifications/handler.py:27]: it's `Maintenance` from the parser library (I added a TODO comment in some place), so I'm waiting for the official release to update

File: nautobot_circuit_maintenance/handle_notifications/handler.py
Comment (chadell) [reply nautobot_circuit_maintenance/handle_notifications/handler.py:256]: because this is for `RawNotification` I intentionally put outside of the `CircuitMaintenance` manipulation. I guess this should affect the signal, no?

File: nautobot_circuit_maintenance/models.py
Comment (chadell) [reply nautobot_circuit_maintenance/models.py:19]: The `CircuitMaintenance` doesn't store the email, only the `RawNotification`, so here we are actually truncating the original notification in the DB, but I believe that we should add some limit to prevent against big email attachements we could receive.
Maybe we could define a lower level to get some minimum size?

File: nautobot_circuit_maintenance/handle_notifications/handler.py
Comment (glennmatthews) [reply nautobot_circuit_maintenance/handle_notifications/handler.py:256]: Ah, my mistake! Fine with me to update the raw entry separately.

File: nautobot_circuit_maintenance/handle_notifications/handler.py
Comment (glennmatthews) [inline nautobot_circuit_maintenance/handle_notifications/handler.py:30]: I just realized in testing my extension plugin that `CircuitMaintenance.objects.create()` doesn't trigger the `pre_save` or `post_save` signals, since it doesn't actually call `save()`. 😞 Would you be willing to change this to something like:

```python
circuit_maintenance_entry = CircuitMaintenance(
   ...
)
circuit_maintenance_entry.save()
```

so that the save signals are properly triggered as I need them to be?

Comment (chadell) [reply nautobot_circuit_maintenance/handle_notifications/handler.py:30]: sure! it's the best timing 😄

Comment (glennmatthews) [reply nautobot_circuit_maintenance/handle_notifications/handler.py]: Thanks!

File: development/nautobot_config.py
Comment (glennmatthews) [inline development/nautobot_config.py:265]: Should any/all of the changes in this file be reverted?

File: nautobot_circuit_maintenance/handle_notifications/handler.py
Comment (glennmatthews) [inline nautobot_circuit_maintenance/handle_notifications/handler.py:321]:
```suggestion
                message=f"Unexpected exception when retrieving notifications from sources ({notification_sources}): {error}"
```

**Review by glennmatthews — APPROVED** [review]

File: development/nautobot_config.py
Comment (chadell) [reply development/nautobot_config.py:265]: I believe these changes are good for development (I have old emails to check 😄 )

**Pull request merged**
